dpar/src/features/input_layers.rs

```diff
@@ -227,6 +271,89 @@ impl InputVectorizer {
             }
         }
     }
+
+    /// Vectorize a parser state into the given association measure slices.
+    ///
+    /// Add to `non_lookup_slices` the association measure between all parser state addresses
+    /// undergoing attachment.
+    pub fn realize_into_assoc_strengths(
+        &self,
+        state: &ParserState,
+        non_lookup_slices: &mut [f32],
+        attachment_addrs: &[AttachmentAddr],
+    ) {
+        for (addr, non_lookup_slice) in attachment_addrs.iter().zip(non_lookup_slices.iter_mut()) {
+            let addr_head = addr::AddressedValue {
+                address: vec![addr.head],
+                layer: addr::Layer::Token,
+            };
+            let addr_dependent = addr::AddressedValue {
+                address: vec![addr.dependent],
+                layer: addr::Layer::Token,
+            };
+
+            let head = addr_head.get(state);
+            let dependent = addr_dependent.get(state);
+
+            if head.is_some()
```

Comment:
Use a `match` or `if let` expression. Along the lines of:

```
if let  (Some(head), Some(dependent), Some(layer)) = (head, dependent, self.layer_lookups.layer_lookup(Layer::DepRel) {
  // ...
}
```

allows you to check that they are all `Some`, while simultaneously unpacking the wrapped values :).

Reply:
Much nicer!